java/vector/src/test/java/org/apache/arrow/vector/util/TestVectorAppender.java

```diff
@@ -0,0 +1,357 @@
+        assertEquals(i, target.get(i));
+      }
+    }
+  }
+
+  @Test
+  public void testAppendVariableWidthVector() {
+    final int length1 = 10;
+    final int length2 = 5;
+    try (VarCharVector target = new VarCharVector("", allocator);
+         VarCharVector delta = new VarCharVector("", allocator)) {
+
+      target.allocateNew(5, length1);
+      delta.allocateNew(5, length2);
+
+      for (int i = 0; i < length1; i++) {
+        target.setSafe(i, ("a" + i).getBytes());
+      }
+      for (int i = 0; i < length2; i++) {
+        delta.setSafe(i, ("a" + (i + length1)).getBytes());
+      }
+      target.setValueCount(length1);
+      delta.setValueCount(length2);
+
+      VectorAppender appender = new VectorAppender(target);
+      delta.accept(appender, null);
+
+      assertEquals(length1 + length2, target.getValueCount());
+      for (int i = 0; i < target.getValueCount(); i++) {
+        assertEquals("a" + i, new String(target.get(i)));
+      }
+    }
+  }
+
+  private ListVector createListVector(int start, int end, int step) {
+    final int listLength = (end - start) / step;
+
+    ListVector listVector = ListVector.empty("list vector", allocator);
+
+    Types.MinorType type = Types.MinorType.INT;
+    listVector.addOrGetVector(FieldType.nullable(type.getType()));
+
+    listVector.allocateNew();
+
+    IntVector dataVector = (IntVector) listVector.getDataVector();
+
+    // set underlying vectors
+    for (int i = 0; i < end - start; i++) {
+      dataVector.set(i, i + start);
+    }
+    dataVector.setValueCount(end - start);
+
+    // set offset buffer
+    for (int i = 0; i < listLength; i++) {
+      BitVectorHelper.setBit(listVector.getValidityBuffer(), i);
+      listVector.getOffsetBuffer().setInt(i * OFFSET_WIDTH, i * step);
+      listVector.getOffsetBuffer().setInt((i + 1) * OFFSET_WIDTH, (i + 1) * step);
+    }
+    listVector.setLastSet(listLength - 1);
+    listVector.setValueCount(listLength);
+
+    return listVector;
+  }
+
+  @Test
+  public void testAppendListVector() {
+    try (ListVector target = createListVector(0, 10, 2);
+         ListVector delta = createListVector(10, 20, 5)) {
+
+      VectorAppender appender = new VectorAppender(target);
+      delta.accept(appender, null);
+
+      assertEquals(7, target.getValueCount());
+
+      int curValue = 0;
+      for (int i = 0; i < 5; i++) {
+        List<Integer> list = (List<Integer>) target.getObject(i);
+        assertEquals(2, list.size());
+        for (int j = 0; j < list.size(); j++) {
+          assertEquals(curValue++, list.get(j).intValue());
+        }
+      }
+
+      for (int i = 5; i < 7; i++) {
+        List<Integer> list = (List<Integer>) target.getObject(i);
+        assertEquals(5, list.size());
+        for (int j = 0; j < list.size(); j++) {
+          assertEquals(curValue++, list.get(j).intValue());
+        }
+      }
+    }
+  }
+
+  private FixedSizeListVector createFixedSizeListVector(int start, int end, int step) {
```

Comment:
I think this method should be eliminated and a method should be added to the test data populators that can take literal values.  It would make the tests easier to read and understand.

Reply:
Good suggestion. I have rewritten the test cases with the test data populators.